src/hypothesis/extra/pandas/impl.py

```diff
@@ -0,0 +1,617 @@
+    else:
+        prefix = '%s.' % (source,)
+
+    if elements is not None:
+        st.check_strategy(elements, '%selements' % (prefix,))
+    else:
+        with check('dtype is not None'):
+            if dtype is None:
+                raise InvalidArgument((
+                    'At least one of %(prefix)selements or %(prefix)sdtype '
+                    'must be provided.') % {'prefix': prefix})
+
+    with check('is_categorical_dtype'):
+        if is_categorical_dtype(dtype):
+            raise InvalidArgument(
+                '%sdtype is categorical, which is currently unsupported' % (
+                    prefix,
+                ))
+
+    dtype = st.try_convert(np.dtype, dtype, 'dtype')
+
+    if elements is None:
+        elements = npst.from_dtype(dtype)
+    elif dtype is not None:
+        def convert_element(value):
+            name = 'draw(%selements)' % (prefix,)
+            try:
+                return np.array([value], dtype=dtype)[0]
+            except TypeError:
+                raise InvalidArgument(
+                    'Cannot convert %s=%r of type %s to dtype %s' % (
+                        name, value, type(value).__name__, dtype.str
+                    )
+                )
+            except ValueError:
+                raise InvalidArgument(
+                    'Cannot convert %s=%r to type %s' % (
+                        name, value, dtype.str,
+                    )
+                )
+        elements = elements.map(convert_element)
+    assert elements is not None
+
+    return elements, dtype
+
+
+class ValueIndexStrategy(st.SearchStrategy):
+    def __init__(self, elements, dtype, min_size, max_size, unique):
+        super(ValueIndexStrategy, self).__init__()
+        self.elements = elements
+        self.dtype = dtype
+        self.min_size = min_size
+        self.max_size = max_size
+        self.unique = unique
+
+    def do_draw(self, data):
+        result = []
+        seen = set()
+
+        iterator = cu.many(
+            data, min_size=self.min_size, max_size=self.max_size,
+            average_size=(self.min_size + self.max_size) / 2
+        )
+
+        while iterator.more():
+            elt = data.draw(self.elements)
+
+            if self.unique:
+                if elt in seen:
+                    iterator.reject()
+                    continue
+                seen.add(elt)
+            result.append(elt)
+
+        dtype = infer_dtype_if_necessary(
+            dtype=self.dtype, values=result, elements=self.elements,
+            draw=data.draw
+        )
+        return pandas.Index(result, dtype=dtype)
+
+
+DEFAULT_MAX_SIZE = 10
+
+
+@st.cacheable
+@st.defines_strategy
+def range_indexes(min_size=0, max_size=None):
+    st.check_valid_interval(min_size, max_size, 'min_size', 'max_size')
+    if max_size is None:
+        max_size = min_size + DEFAULT_MAX_SIZE
+    return st.integers(min_size, max_size).map(
+        lambda i: pandas.Index(hrange(i), dtype=int)
```

Comment:
This line kept failing for me, with errors about requiring a 64bit integer.  Instead of using hrange, it seems to work well as

      return st.integers(min_size, max_size).map(np.arange).map(pandas.Index)

Reply:
Interesting. I've never seen that failure. What sort of args does it fail on?

Reply:
I've added a test trying arbitrary arguments to range_indexes and I've *still* never seen it fail. I'm happy to make the change, but if there's a bug lurking here I'd like to know what it is.

Reply:
I was running it in a notebook, literally without arguments, and it failed.  I've just tried it again on the latest version of this branch with the same result - but still no idea why.

    pdst.range_indexes().example()

<details>
  <summary>Full traceback</summary>

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-276c872cfbaa> in <module>()
----> 1 pdst.range_indexes().example()

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\searchstrategy\strategies.py in example(self, random)
    179                     max_iterations=1000,
    180                     database=None,
--> 181                     verbosity=Verbosity.quiet,
    182                 )
    183             )

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\core.py in find(specifier, condition, settings, random, database_key)
    848         database_key=database_key,
    849     )
--> 850     runner.run()
    851     note_engine_for_statistics(runner)
    852     run_time = time.time() - start

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\internal\conjecture\engine.py in run(self)
    364         with self.settings:
    365             try:
--> 366                 self._run()
    367             except RunIsComplete:
    368                 pass

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\internal\conjecture\engine.py in _run(self)
    635                 self.last_data.status < Status.INTERESTING
    636             ):
--> 637                 self.new_buffer()
    638 
    639             mutator = self._new_mutator()

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\internal\conjecture\engine.py in new_buffer(self)
    120             draw_bytes=draw_bytes
    121         )
--> 122         self.test_function(self.last_data)
    123         self.last_data.freeze()
    124 

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\internal\conjecture\engine.py in test_function(self, data)
    126         self.call_count += 1
    127         try:
--> 128             self._test_function(data)
    129             data.freeze()
    130         except StopTest as e:

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\core.py in template_condition(data)
    817             try:
    818                 data.is_find = True
--> 819                 result = data.draw(search)
    820                 data.note(result)
    821                 success = condition(result)

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\internal\conjecture\data.py in draw(self, strategy)
    111         self.start_example()
    112         try:
--> 113             return strategy.do_draw(self)
    114         finally:
    115             if not self.frozen:

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\searchstrategy\lazy.py in do_draw(self, data)
    152 
    153     def do_draw(self, data):
--> 154         return data.draw(self.wrapped_strategy)

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\internal\conjecture\data.py in draw(self, strategy)
    111         self.start_example()
    112         try:
--> 113             return strategy.do_draw(self)
    114         finally:
    115             if not self.frozen:

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\searchstrategy\strategies.py in do_draw(self, data)
    389             i = data.index
    390             try:
--> 391                 return self.pack(data.draw(self.mapped_strategy))
    392             except UnsatisfiedAssumption:
    393                 if data.index == i:

c:\users\zac\documents\github\hypothesis-python\src\hypothesis\extra\pandas\impl.py in <lambda>(i)
    161         max_size = min_size + DEFAULT_MAX_SIZE
    162     return st.integers(min_size, max_size).map(
--> 163         lambda i: pandas.Index(hrange(i), dtype=int)
    164     )
    165 

C:\Anaconda3\envs\hydev\lib\site-packages\pandas\core\indexes\base.py in __new__(cls, data, dtype, copy, name, fastpath, tupleize_cols, **kwargs)
    166         elif isinstance(data, range):
    167             return RangeIndex.from_range(data, copy=copy, dtype=dtype,
--> 168                                          name=name)
    169 
    170         # categorical

C:\Anaconda3\envs\hydev\lib\site-packages\pandas\core\indexes\range.py in from_range(cls, data, name, dtype, **kwargs)
    121                 start = stop = 0
    122                 step = 1
--> 123         return RangeIndex(start, stop, step, dtype=dtype, name=name, **kwargs)
    124 
    125     @classmethod

C:\Anaconda3\envs\hydev\lib\site-packages\pandas\core\indexes\range.py in __new__(cls, start, stop, step, name, dtype, fastpath, copy, **kwargs)
     49             return cls._simple_new(start, stop, step, name=name)
     50 
---> 51         cls._validate_dtype(dtype)
     52 
     53         # RangeIndex

C:\Anaconda3\envs\hydev\lib\site-packages\pandas\core\indexes\range.py in _validate_dtype(dtype)
    154         """ require dtype to be None or int64 """
    155         if not (dtype is None or is_int64_dtype(dtype)):
--> 156             raise TypeError('Invalid to pass a non-int64 dtype to RangeIndex')
    157 
    158     @cache_readonly

TypeError: Invalid to pass a non-int64 dtype to RangeIndex
```

</details>

Reply:
What pandas/python versions?

Reply:
I actually dumped this at the same time but forgot to copy it in after I looked up how to do a collapsed section 😊   It's basically an empty environment with an editable Hypothesis dev install, and the latest versions of Jupyter, Numpy, and Pandas on conda.

<details>
  <summary>pd.show_versions()</summary>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 8.1
machine: AMD64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.3
pytest: 3.1.1
pip: 9.0.1
setuptools: 36.0.1
Cython: None
numpy: 1.13.1
scipy: None
xarray: None
IPython: 6.0.0
sphinx: 1.6.2
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
```

</details>

Reply:
This is really odd. We've got a whole bunch of tests running this code and none of them are exhibiting this failure. Maybe it's a pandas on windows thing?

Hmph. We probably *should* be testing pandas on windows, but our appveyor builds are slow enough as it is.

Anyway, I'll update the code to just create a RangeIndex directly, but this worries me.

Reply:
Also worried here, but I don't see a good way to deal with it.  I'll poke at it more when using this on a toy project though.

> Hmph. We probably should be testing pandas on windows, but our appveyor builds are slow enough as it is.

Maybe try asking them for free concurrent builds?  There's a decent case that supporting Hypothesis will lead to higher use of their platform, after all, and asking is cheap 😄